New fusion graph with explicit buffers in preparation for in-place updates

BlueFlame-SM opened the pull request:
For in-place updates we require the presence of buffers in the graph used by the ILP.
The new graph construction, solving, and reconstruction of ASTs should now be ready.

Due to the new graph structure it was necessary to rewrite almost all of the graph construction code, however solving and reconstruction remained relatively the same.
There are some small changes here, in that the numberOfReads cost-function has now been massively simplified(?).
In addition to that there where some small changes to the constraints to ensure the new split between strict and dataflow edges works.
As for the reconstruction of ASTs, it used to be that the backend stored information it needed for reconstruction on the arguments to an execute. This is no longer required with this new system, as it can extract this information directly from the solution now. This may have made it so BackendArg is now redundant, we need to check.

BlueFlame-SM committed:
Update haskell version

BlueFlame-SM committed:
Merge branch 'new-pipeline' of https://github.com/ivogabe/accelerate into new-pipeline

BlueFlame-SM committed:
Ignore /.vscode

BlueFlame-SM committed:
Merge branch 'new-pipeline' of https://github.com/ivogabe/accelerate into new-pipeline

BlueFlame-SM committed:
Init additional files

BlueFlame-SM committed:
Created all cases for mkDataFlowGraph

BlueFlame-SM committed:
Labels

BlueFlame-SM committed:
Setup for writing new graph builder

BlueFlame-SM committed:
Logic for creating edges in 'exec'

BlueFlame-SM committed:
Execute working?

BlueFlame-SM committed:
Most cases done.

BlueFlame-SM committed:
If-then-else preparations

BlueFlame-SM committed:
todo: afun

BlueFlame-SM committed:
Ready for Afun

BlueFlame-SM committed:
implementation with result family

BlueFlame-SM committed:
commit

BlueFlame-SM committed:
Merge branch 'new-pipeline' of https://github.com/ivogabe/accelerate into new-pipeline

BlueFlame-SM committed:
commit

BlueFlame-SM committed:
Added simple version of backend specific definitions

BlueFlame-SM committed:
Intermediate commit so git knows I replaced these with the new version

BlueFlame-SM committed:
Replaced old graph implementation with new

BlueFlame-SM committed:
changes

BlueFlame-SM committed:
Ignore /.vscode

BlueFlame-SM committed:
tracy

BlueFlame-SM committed:
Solved merge issue where partitioned.hs relied on functionality of old labels implementation.
I.e. re-added this old functionality with a new flavor.

BlueFlame-SM committed:
More errors fixed

BlueFlame-SM committed:
Idk what these changes are but sure

BlueFlame-SM committed:
Graph construction now produces multiple sub-ilp's

BlueFlame-SM committed:
Back to full graph again, by adding bounds

BlueFlame-SM committed:
Very small refactor

BlueFlame-SM committed:
Lambdas now also infusibly attach to results

BlueFlame-SM committed:
Resolved the errors in clustering.hs

BlueFlame-SM committed:
I think mostly done now

BlueFlame-SM committed:
Added back backendConstruc and renamed it to attachBackendLabels

BlueFlame-SM committed:
Disabled greedyFusion' completely (couldn't be bothered fixing it with the right types for now)

BlueFlame-SM committed:
Compiles without errors

BlueFlame-SM committed:
Removed Interpreter from the cabal file

BlueFlame-SM committed:
Some style changes

BlueFlame-SM committed:
Another small change

BlueFlame-SM committed:
Commented out only unnecessary parts of interpreter.hs (backend imports some stuff from it)

BlueFlame-SM committed:
Formatted the Language pragmas and imports + removed temporary Edge

BlueFlame-SM committed:
Restricted the backend to a small environment so it can't accidentally mess up the graph construction process.

BlueFlame-SM committed:
Added back manual definition of read and write edges, since the prior knowledge of a write relation before a computation reads from a buffer might be useful.

BlueFlame-SM committed:
Everything should typecheck.

BlueFlame-SM committed:
Fixed a bug where arguments that are both In and Out (but not Mut) would insert a reflexive edge. Switched to strict State for debugging.

BlueFlame-SM committed:
Added a very simple DOT graph renderer

BlueFlame-SM committed:
Trying to figure out missing lets

BlueFlame-SM committed:
changes

BlueFlame-SM committed:
ITS ALIVE!

BlueFlame-SM committed:
Solved an issue where fusion was blocked by the shape variable.

BlueFlame-SM committed:
Fixed an issue where the same operation tried writing to the same buffer multiple times, thus trying to be infusible with itself.
Fixed an issue where we did not create all required read edges.

BlueFlame-SM committed:
Fixed an issue with bodies of if-then-else and while

BlueFlame-SM committed:
Changed interpretSolution to be more readable (and now I understand what it does)

BlueFlame-SM committed:
Added some functions for debugging

BlueFlame-SM committed:
Fixed an issue where while-loops returned the buffer a function writes to instead of the input arrays.

BlueFlame-SM committed:
Still works if I don't do stupid and forget the symbolic mutate

BlueFlame-SM committed:
Preparation for rendering large MG graph

BlueFlame-SM committed:
Fixed an issue where some let bindings wouldn't have a body? Maybe?

BlueFlame-SM committed:
Some style changes

BlueFlame-SM committed:
Fixed an issue where the reconstruction could not locate a read direction. BackendArgs are now redundant.

BlueFlame-SM committed:
Reverted removal of old "numberOfReads" code.

BlueFlame-SM committed:
removed traceWith code

BlueFlame-SM committed:
Better solution pretty printing

BlueFlame-SM committed:
Added a comment clarifieing that EnvLabels may now be redundant.

BlueFlame-SM committed:
result of awhile now contains the initial value

ivogabe commented on src/Data/Array/Accelerate/Trafo/Partitioning/ILP/Labels.hs:
```diff
@@ -1,340 +1,641 @@
+--   left-hand side.
+--
+-- The case where the left-hand side and the right-hand side are incompatible
+-- should neven happen, but in case it does just replicate the labels.
```
neven -> never. Is this comment still correct? You throw an error when the LHS is a pair and the RHS not (which makes sense to me).

ivogabe commented on src/Data/Array/Accelerate/Trafo/Partitioning/ILP/Labels.hs:
```diff
@@ -1,340 +1,641 @@
+weakenEnv LeftHandSideWildcard{} _                  = pure
+weakenEnv LeftHandSideSingle{}   bs                 = \lenv -> freshE' >>= \e -> return ((e, bs) :>>: lenv)
+weakenEnv (LeftHandSidePair l r) (TupFpair lbs rbs) = weakenEnv l lbs >=> weakenEnv r rbs
+weakenEnv (LeftHandSidePair _ _) _ = error "consLHS: Inaccesible left-hand side."
```
error -> internalError

ivogabe commented on src/Data/Array/Accelerate/Trafo/Partitioning/ILP/Graph.hs:
```diff
@@ -1,125 +1,509 @@
+-- because the size variable is required by the allocator, it is still enforced
+-- implicitly.
+--
+data FusionGraph = FusionGraph   -- TODO: Use hashmaps and hashsets in production.
```
Do you mean that you want to use your Multimap here? I think that makes sense, but maybe we can also have another look whether Sets would be sufficient. For instance, with lookupGT on a set of tuples you can emulate a lookup in a map.

ivogabe commented on src/Data/Array/Accelerate/Trafo/Partitioning/ILP/Graph.hs:
```diff
@@ -1,125 +1,509 @@
+-- @
+--
+-- Finally, I've added size edges to the graph. These edges point from a buffer
+-- storing an array to the buffer storing its size. We do this because we need
```
What do you mean with 'the buffer storing its size'? A size of a buffer is stored in a scalar variable, not a buffer. Sometimes that variable is defined by reading from a buffer (`let n = b ! 0`), but you're probably not referring to that scenario.

ivogabe commented on src/Data/Array/Accelerate/Trafo/Partitioning/ILP/Graph.hs:
```diff
@@ -1,125 +1,509 @@
+-- | Insert a strict relation between two computations.
+insertStrict :: (HasCallStack, HasFusionGraph g) => StrictEdge -> g -> g
+insertStrict (c1, c2) g
+  | c1 == c2                           = error $ "insertStrict " ++ show (c1, c2) ++ ": Reflexive edge"
```
Also here: use internalError (from Data.Array.Accelerate.Error) instead of error

ivogabe commented on src/Data/Array/Accelerate/Trafo/Partitioning/ILP/Labels.hs:
```diff
@@ -1,340 +1,641 @@
+--------------------------------------------------------------------------------
+
+-- | The types a label can have.
+data LabelType = Comp | Buff
```
Are you using Buff both for Buffer values and scalar values? It seems like that at least. Since a GroundR is either a buffer or a scalar, you will have variables of those two types. Either change Buff to Value (easiest fix) or add a Scalar constructor (if you need to separate labels to Buffers from labels to scalars)

ivogabe commented on src/Data/Array/Accelerate/Trafo/Desugar.hs:
```diff
@@ -333,7 +333,7 @@ class NFData' op => DesugarAcc (op :: Type -> Type) where
   -- be converted to Compute nodes, instead of Exec nodes.
   desugarPreferNoScalar :: Bool
-  desugarPreferNoScalar = True
+  desugarPreferNoScalar = False
```
Change this back to True (this was for testing I think, I remember suggesting this change to test something)

ivogabe commented on src/Data/Array/Accelerate/Trafo/Partitioning/ILP/Clustering.hs:
```diff
@@ -55,25 +58,28 @@ import Data.String (fromString)
 -- -- Some of those ints will refer to nodes in previous clusters, others to nodes in this cluster.
 -- One pass over these datatypes (back-to-front) should identify the 'output type' of each cluster: which nodes are needed in later clusters?
--- Then, we can construct the clusters front-to-back:
+-- Then, we can symbols the clusters front-to-back:
```
Was this change intended? (Same for a few lines below)

BlueFlame-SM committed:
Resolved the suggesionts made by Ivo (except Buff -> Value).

BlueFlame-SM committed:
Applied same let-binding blocking non-manifest array fix as on inplace-updates branch.

BlueFlame-SM committed:
Merge branch 'new-pipeline' of https://github.com/ivogabe/accelerate into new-ilp

Note: Can't seem to compile nofib-interpreter test-suite (variable runN) not in scope.

BlueFlame-SM committed:
Deleted Multimap.hs

BlueFlame-SM committed:
Merge branch 'new-pipeline' of github.com:ivogabe/accelerate into new-pipeline

BlueFlame-SM committed:
Fixed HLS not working on some files.
+ Added missing declaration for memoryByteSize in __GHCIDE__ preprocessor branch.

BlueFlame-SM committed:
Revert "tracy"

This reverts commit 980a8628f7794b6f4bad9ff8a4c2c9c5b59b9402.

BlueFlame-SM committed:
Merge branch 'new-pipeline' into new-ilp

BlueFlame-SM committed:
Set test-suite to use simple interpreter

BlueFlame-SM committed:
Removed duplicated lines

ivogabe commented:
As discussed offline, this PR will be superseded by a new PR with both the new fusion system and in-place updates
